### ISSUE TITLE
Model outputs zero detections

### ISSUE THREAD

**Issue by fernandocamargoti** [fizyr/keras-retinanet#694]
I've been training with a custom dataset of X-Rays to detect pneumonia and the loss is reducing during traning. But when I create the retinanet_bbox() model and try to evaluate it, it doesn't output any detection. I've debugged the moment it invokes predict_on_batch() and it's returning only -1 scores, -1 labels and [-1 -1 -1 -1] boxes, which is turned into zero detections. I even tried to use 0.0 score_threshold, but it doesn't  matter.

**Comment by hgaiser** [comment]
Most likely the problem is in your input, try running evaluate.py with
--save-path and --score-threshold 0.5

**Comment by fernandocamargoti** [comment]
I've used save-path and looked at the images. The images look normal and the bounding boxes are on the right places. But only the ground truth bounding boxes are being drawn in the image. No predictions are being drawn.

The loss and val_loss are both around 1.7. If the regression loss is decreasing, it should be output some bounding boxes, I guess. But when I try to evaluate, there's nothing in the output.

**Comment by hgaiser** [comment]
1.7 is quite high, in that case I assume you're training data is somehow not right.

**Comment by fernandocamargoti** [comment]
I'm training with the data available from this competition:
https://www.kaggle.com/c/rsna-pneumonia-detection-challenge

The problem is really hard to solve. Maybe that's why the loss is very high yet. But what I don't understand is how the loss is decreasing if no bounding boxes are being generated. I'd expect to have bounding boxes off the target, but not having zero bounding boxes.

**Comment by fernandocamargoti** [comment]
Could be something related to the image sizes? They're originally 1024x1024. I'm resizing them to 512x512 (I'm resizing the bounding boxes as well).

**Comment by fernandocamargoti** [comment]
To be sure that it wasn't about the loss value, I removed the validation and all the regularization and let it training despite the overfitting. Right now, the loss is 0.27 and still no detection is generated.

**Comment by fernandocamargoti** [comment]
I discovered the problem. I have only one class and was returning 1 in num_class(). But my class id was 1 instead of 0. It seems that it really needs to start at 0, otherwise the class is ignored. I know it was my fault, but I guess it'd be very useful to have some kind of warning whenever an invalid class id is found.

**Comment by de-vri-es** [comment]
I take it you implemented a custom data generator? I'm not so sure we can really inspect the data after it leaves the generator but before it is fed to the network.

At best the data generator itself could check the output, but that would require an opt-in from the data generator. Or we would have to wrap all generators in another generator that checks the output, which might have a performance impact.

I suppose we could check for this type of error in debug.py and evaluate.py though. That could be useful. @hgaiser: Thoughts?

**Comment by hgaiser** [comment]
Anyway I guess this issue can be closed.
